asset-transfer-secured-agreement/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,427 @@
+        await changePublicDescriptionAfterTransfer(assetKey,mspIdOrg2,`Asset ${assetKey} owned by ${mspIdOrg2} is NOT for sale`,contractOrg2);
+
+        // Read the public details by  org1.
+        await readAsset(assetKey, mspIdOrg2, contractOrg1,mspIdOrg1);
+
+        // Read the public details by  org2.
+        await readAsset(assetKey, mspIdOrg2, contractOrg2,mspIdOrg2);
+
+    } finally {
+        gatewayOrg1.close();
+        gatewayOrg2.close();
+        clientOrg1.close();
+        clientOrg2.close();
+    }
+}
+
+main().catch(error => {
+    console.error('******** FAILED to run the application:', error);
+    process.exitCode = 1;
+});
+
+async function createAsset(contract:Contract,org:string):Promise<void> {
+    const asset_properties = {
+        object_type: 'asset_properties',
+        asset_id: assetKey,
+        color: 'blue',
+        size: 35,
+        salt: Buffer.from(now.toString()).toString('hex')
+    };
+    console.log(`${GREEN}--> Submit Transaction: CreateAsset, ${assetKey} as ${org} - endorsed by Org1${RESET}`);
+
+    await contract.submit('CreateAsset', {
+        arguments:[assetKey,`Asset ${assetKey} owned by ${org} is not for sale`],
+        transientData: { asset_properties: JSON.stringify(asset_properties) },
+    });
+
+    console.log(`*** Result: committed, asset ${assetKey} is owned by Org1`);
+}
+
+async function readAsset(assetKey:string, ownerOrg:string, contract:Contract,org:string) {
+    console.log(`${GREEN}--> Evaluate Transactions: ReadAsset as ${org}, - ${assetKey} should be owned by ${ownerOrg}${RESET}`);
+
+    const resultBytes = await contract.evaluateTransaction('ReadAsset', assetKey);
+
+    const resultString = utf8Decoder.decode(resultBytes);
+    const result = JSON.parse(resultString);
+
+    if (result) {
+        if (result.ownerOrg === ownerOrg) {
+            console.log(`*** Result from ${org} - asset ${result.assetID} owned by ${result.ownerOrg} DESC:${result.publicDescription}`);
+        } else {
+            console.log(`${RED}*** Failed owner check from ${org} - asset ${result.assetID} owned by ${result.ownerOrg} DESC:${result.publicDescription}${RESET}`);
+        }
+    }
+}
+
+async function readPrivateAsset(assetKey:string, org:string, contract:Contract):Promise<void> {
+    try{
+        console.log(`${GREEN}--> Evaluate Transaction: GetAssetPrivateProperties, - ${assetKey} from organization ${org}${RESET}`);
+        if(org === mspIdOrg2){
+            console.log(`${GREEN}* Expected to fail as ${org} is not the owner and does not have the private details${RESET}`)
+        }
+        const resultBytes = await contract.evaluateTransaction('GetAssetPrivateProperties', assetKey);
+
+        const resultString = utf8Decoder.decode(resultBytes);
+        const result = JSON.parse(resultString);
+
+        console.log('*** Result:', result);
+    }
+    catch(e){
+        console.log(`${RED}*** Failed evaluateTransaction readPrivateAsset: ${e}${RESET}`);
```

Comment:
Should the error be sent back to the application here?  As far as the app is concerned all the transactions will execute successfully?

Reply:
The transaction is expected to fail .This is to demonstrate access control as org2 is not the owner and does not have the private details.